File: common/src/main/java/juuxel/adorn/networking/BrewerFluidSyncS2CMessage.java
package juuxel.adorn.networking;

import juuxel.adorn.AdornCommon;
import juuxel.adorn.fluid.FluidVolume;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;

public record BrewerFluidSyncS2CMessage(int syncId, FluidVolume fluid) implements CustomPacketPayload {
    public static final Type<BrewerFluidSyncS2CMessage> ID = new Type<>(AdornCommon.id("brewer_fluid_sync"));
    public static final StreamCodec<RegistryFriendlyByteBuf, BrewerFluidSyncS2CMessage> PACKET_CODEC = StreamCodec.composite(
        ByteBufCodecs.VAR_INT,
        BrewerFluidSyncS2CMessage::syncId,
        FluidVolume.PACKET_CODEC,
        BrewerFluidSyncS2CMessage::fluid,
        BrewerFluidSyncS2CMessage::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return ID;
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/networking/BrewerFluidSyncS2CMessage.kt
package juuxel.adorn.platform.forge.networking

import juuxel.adorn.fluid.FluidVolume
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.packet.CustomPayload
import net.minecraft.util.Identifier

data class BrewerFluidSyncS2CMessage(val syncId: Int, val fluid: FluidVolume) : CustomPayload {
    override fun id(): Identifier = AdornNetworking.BREWER_FLUID_SYNC

    override fun write(buf: PacketByteBuf) {
        buf.writeByte(syncId)
        fluid.write(buf)
    }

    companion object {
        fun fromPacket(buf: PacketByteBuf): BrewerFluidSyncS2CMessage =
            BrewerFluidSyncS2CMessage(
                syncId = buf.readUnsignedByte().toInt(),
                fluid = FluidVolume.load(buf)
            )
    }
}
